toolchain/format/formatter.cpp

```diff
@@ -0,0 +1,110 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/format/formatter.h"
+
+namespace Carbon::Format {
+
+auto Formatter::Run() -> bool {
+  if (tokens_->has_errors()) {
+    // TODO: Error recovery.
+    return false;
+  }
+
+  auto comments = tokens_->comments();
+  auto comment_it = comments.begin();
+
+  // If there are no tokens or comments, format as empty.
+  if (tokens_->size() == 0 && comment_it == comments.end()) {
+    *out_ << "\n";
+    return true;
+  }
+
+  for (auto token : tokens_->tokens()) {
+    auto token_kind = tokens_->GetKind(token);
+
+    while (comment_it != comments.end() &&
+           tokens_->IsAfterComment(token, *comment_it)) {
+      AddNewline();
+      AddWhitespace();
+      // TODO: We do need to adjust the indent of multi-line comments.
+      *out_ << tokens_->GetCommentText(*comment_it);
+      // Comment text includes a terminating newline, so just update the state.
+      line_state_ = LineState::Empty;
+      ++comment_it;
+    }
+
+    switch (token_kind) {
+      case Lex::TokenKind::FileStart:
+        break;
+
+      case Lex::TokenKind::FileEnd:
+        AddNewline();
+        break;
+
+      case Lex::TokenKind::OpenCurlyBrace:
+        AddWhitespace();
+        *out_ << "{";
```

Comment:
It seems like this should be followed by `line_state_ = LineState::HasContent;`, or at least a comment indicating that we're intentionally carrying over the `HasContent` state from `AddWhitespace`, and the same goes for the other places where we write characters to the output without modifying `line_state_`. That would make the state-machine structure clearer: we write some text to the current line, and then we update `line_state_` to keep it in sync with the actual state of the line.

Reply:
It sounds like you're suggesting documenting function semantics at each call site, instead of the function itself. I think it's better to aim for central, rather than duplicated, documentation.

How about this: I've adjusted function names to try to better reflect semantics: RequireEmptyLine/PrepareForSpacedContent/PrepareForPackedContent. Does that work for you?

Reply:
By the way, if you want the explicit notes that line_state is modified, I can always revert back to the original (non-class) structure.